code/datums/helper_datums/teleport.dm

```diff
@@ -1,13 +1,80 @@
-// no_effects: disable the default effectin/effectout of sparks
-// forced: whether or not to ignore no_teleport
-/proc/do_teleport(atom/movable/teleatom, atom/destination, precision=null, datum/effect_system/effectin=null, datum/effect_system/effectout=null, asoundin=null, asoundout=null, no_effects=FALSE, channel=TELEPORT_CHANNEL_BLUESPACE, forced = FALSE, teleport_mode = TELEPORT_MODE_DEFAULT, commit = TRUE, no_wake = FALSE)
+/**
+ * Returns FALSE if we SHOULDN'T do_teleport() with the given arguments
+ *
+ * Arguments:
+ * * teleatom: The atom to teleport
+ * * destination: The destination turf for the atom to go
+ * * precision: How accurate should the teleport be (in tiles), defaults to 0
+ * * channel: Which teleport channel should we use (for checks), defaults to TELEPORT_CHANNEL_BLUESPACE
+ * * forced: Do we ignore atom and area TRAIT_NO_TELEPORT restrictions, defaults to FALSE
+ * * teleport_mode: Teleport mode which allows ONLY clockies and abductors to teleport in their area
+ */
+/proc/check_teleport(atom/movable/teleatom, turf/dest_turf, channel = TELEPORT_CHANNEL_BLUESPACE, forced = FALSE, teleport_mode = TELEPORT_MODE_DEFAULT)
+	var/turf/cur_turf = get_turf(teleatom)
+
+	if(!cur_turf || !dest_turf || dest_turf.is_transition_turf())
+		return FALSE
+
+	// Check for bluespace anchors
+	if(channel != TELEPORT_CHANNEL_FREE && channel != TELEPORT_CHANNEL_WORMHOLE && channel != TELEPORT_CHANNEL_GATEWAY)
+		for (var/obj/machinery/bluespace_anchor/anchor as() in GLOB.active_bluespace_anchors)
+			var/cur_zlevel = anchor.get_virtual_z_level()
+			// Current not in range
+			if(cur_zlevel != cur_turf.get_virtual_z_level() && get_dist(cur_turf, anchor) > anchor.range)
+				continue
+			// Destination not in range
+			if(cur_zlevel != dest_turf.get_virtual_z_level() && get_dist(dest_turf, anchor) > anchor.range)
+				continue
+			// Try to activate the anchor, this also does the effect
+			if(!anchor.try_activate())
+				continue
+			// We're anchored, return false
+			return FALSE
+
+	// Check for NO_TELEPORT restrictions
+	if(!forced)
+		var/area/cur_area = cur_turf.loc
+		var/area/dest_area = dest_turf.loc
+		if(HAS_TRAIT(teleatom, TRAIT_NO_TELEPORT))
+			return FALSE
+		if(cur_area.teleport_restriction && cur_area.teleport_restriction != teleport_mode)
+			return FALSE
+		if(dest_area.teleport_restriction && dest_area.teleport_restriction != teleport_mode)
+			return FALSE
+
+	// Check for intercepting the teleport
+	if(SEND_SIGNAL(cur_turf, COMSIG_ATOM_INTERCEPT_TELEPORT, channel, cur_turf, dest_turf))
+		return FALSE
+	if(SEND_SIGNAL(dest_turf, COMSIG_ATOM_INTERCEPT_TELEPORT, channel, cur_turf, dest_turf))
+		return FALSE
+	if(SEND_SIGNAL(teleatom, COMSIG_ATOM_INTERCEPT_TELEPORT, channel, cur_turf, dest_turf))
+		return FALSE
+
+	// Recursively check contents to prevent cheese
+	for(var/atom/thing in teleatom.GetAllContents())
+		if(SEND_SIGNAL(thing, COMSIG_ATOM_INTERCEPT_TELEPORT, channel, cur_turf, dest_turf))
+			return FALSE
```

Comment:
I find it much more easier especially in patching any unintended behavior where otherwise we'll need to register teleport interception for all storage objects, all mobs, etc. This takes care of that in a way that doesn't add more LOC and ` is probably more optimized than recursively calling `COMSIG_ATOM_INTERCEPT_TELEPORT` on every atom that a mob or crate for example contains.

Reply:
Anyways to implement the intended feature of preventing any sort of edge-case where we forgot to implement a recursive INTERCEPT_TELEPORT on a new type of mob (*cough* basic mob) or a new special item for example.